src/payment/strategies/zip/zip-payment-strategy.ts

```diff
@@ -0,0 +1,159 @@
+        }
+
+        return this._zipScriptLoader.load()
+            .then(zip => {
+                this._zipClient = zip;
+
+                return this._store.getState();
+            });
+    }
+
+    deinitialize(options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        this._paymentMethod = undefined;
+        this._zipClient = undefined;
+
+        return Promise.resolve(this._store.getState());
+    }
+
+    execute(payload: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        if (!this._paymentMethod) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        return this._displayZip(payload);
+    }
+
+    finalize(options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        return Promise.reject(new OrderFinalizationNotRequiredError());
+    }
+
+    private _displayZip(payload: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        return this._store.dispatch(this._paymentStrategyActionCreator.widgetInteraction(() => {
+            this._store.dispatch(this._paymentMethodActionCreator.loadPaymentMethod(this._methodId))
+                .then(state => {
+                    if (!this._zipClient) {
+                        throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
+                    }
+
+                    this._zipClient.Checkout.init(this._getZipPayload());
+                });
+
+            // Lightbox closing event listener
+            return new Promise((resolve, reject) => {
+                this._lightboxEvents$
+                    .pipe(take(1))
+                    .subscribe((event: { type: ZipModalEvent }) => {
+                        if (event.type === ZipModalEvent.CancelCheckout) {
+                            reject(new PaymentMethodCancelledError());
+                        }
+                        if (event.type === ZipModalEvent.CheckoutApproved) {
+                            resolve(this._submitOrder(payload, options));
+                        }
+                    });
+            });
+        }, { methodId: this._methodId }), { queueId: 'widgetInteraction' });
+    }
+
+    private _getZipPayload(): ZipPayload {
+        return {
+            onComplete: this._handleResponse,
+            onCheckout: this._fetchPayload,
+        };
+    }
+
+    private _submitOrder(payload: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        const order = { useStoreCredit: payload.useStoreCredit };
+
+        if (!this._zipLightboxResponse) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        const paymentPayload = {
+            methodId: this._methodId,
+            paymentData: { nonce: this._zipLightboxResponse.checkoutId } as NonceInstrument,
```

Comment:
Under normal circumstances, at this point it should not be undefined, but `checkoutId` is defined as `string?` because the Lightbox response can be `state: 'cancelled'` only and completely omit the `checkoutId` on that case.

Reply:
I see, you're using the same interface for both success and cancelled state. I guess can be worked around by having a simple `if` statement checking the presence of `checkoutId`.